synapse/handlers/saml_handler.py

```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
+
+            redirect_url = next(
+                header[1] for header in http_info["headers"] if header[0] == "Location"
+            )
+            if not redirect_url:
+                raise RuntimeError("missing Location header")
+
+            return redirect_url
+
+        except Exception as e:
+            raise SynapseError(
+                500, "error while creating SAML logout request: %s" % (e,)
+            )
+
+    async def handle_logout_request(self, request):
+        """Handle an incoming LogoutRequest to /_matrix/saml2/logout
+
+        Args:
+            request (bytes): a SAML LogoutRequest
+
+        Returns:
+            bytes: URL to redirect to
+        """
+        saml_req_encoded = parse_string(request, "SAMLRequest", required=True)
+        relay_state = parse_string(request, "RelayState")
+        # TODO: sign LogoutRequest responses if required by the IdP
+        # sign = parse_string(request, "SigAlg")
+        # sign_alg = parse_string(request, "Signature")
+
+        saml_req = self._saml_client.parse_logout_request(
+            saml_req_encoded, BINDING_HTTP_REDIRECT
+        )
+        name_id = saml_req.message.name_id
+        mxid = self._find_mxid_from_name_id(name_id)
+
+        # Logout from matrix
+        if mxid:
+            await self._logout(mxid)
+
+        # Logout from the local SAML cache
+        try:
+
+            if self._saml_client.local_logout(name_id):
+                status = success_status_factory()
+            else:
+                status = status_message_factory("Server error", STATUS_REQUEST_DENIED)
+        except KeyError:
+            status = status_message_factory("Server error", STATUS_REQUEST_DENIED)
+
+        # Prepare SAML LogoutResponse using HTTP_REDIRECT
+        response = self._saml_client.create_logout_response(
+            saml_req.message, [BINDING_HTTP_REDIRECT], status
+        )
+        rinfo = self._saml_client.response_args(
+            saml_req.message, [BINDING_HTTP_REDIRECT]
+        )
+
+        rfinal = self._saml_client.apply_binding(
+            rinfo["binding"], response, rinfo["destination"], relay_state, response=True
+        )
+
+        # Return the redirect_url
+        for key, value in rfinal["headers"]:
+            if key == "Location":
+                return value
+
+        # this shouldn't happen!
+        raise Exception("create_logout_response didn't return a Location header")
+
+    def handle_logout_response(self, request):
+        """
+            Handle an incoming LogoutResponse to /_matrix/saml2/logout
+        """
+        resp_bytes = parse_string(request, "SAMLResponse", required=True)
+        try:
+            resp_saml = self._saml_client.parse_logout_request_response(
+                resp_bytes, BINDING_HTTP_REDIRECT
+            )
+            logger.info("Received SAML logout response %s", resp_saml)
+            if resp_saml.response.status.status_code.value == STATUS_SUCCESS:
```

Comment:
suggest inverting this condition

Reply:
what do you mean ?
if STATUS_SUCCESS == resp_saml.response.status.status_code.value: ?